
# âš”ï¸  Engineering Rules

> **Version**: 3.0.0 (Monorepo Standard)
> **Enforcement**: Strict. Violations will result in rejected PRs.

## 1. ğŸ—ï¸ æŠ€æœ¯æ ˆæ¸…å• (Tech Stack)

æ‰€æœ‰å¼€å‘æ´»åŠ¨å¿…é¡»é™åˆ¶åœ¨ä»¥ä¸‹æŠ€æœ¯èŒƒå›´å†…ã€‚

| é¢†åŸŸ | æ ¸å¿ƒæŠ€æœ¯ | å…³é”®åº“/å·¥å…· | ç‰ˆæœ¬/å¤‡æ³¨ |
| :--- | :--- | :--- | :--- |
| **Monorepo** | **Turborepo** | `pnpm` (Workspace) | **å¼ºåˆ¶ pnpm** |
| **Language** | **TypeScript** | Strict Mode | v5.0+ |
| **Frontend** | **React** | Vite, React Router 6 | v18+ |
| **UI System** | **Ant Design 5** | **ProComponents** (Table/Form) | Token-based styling |
| **State (FE)** | **TanStack Query** | Zustand (Client Global) | v5+ |
| **Backend** | **NestJS** | Express, RxJS | v10+ |
| **Database** | **PostgreSQL** | **Prisma ORM** | Latest |
| **Validation** | **Zod** | `zod`, `nestjs-zod` | Unified Schema |

---

## 2. ğŸ—ºï¸ é¡¹ç›®ç›®å½•ç»“æ„ (Directory Structure)

æˆ‘ä»¬é‡‡ç”¨ **Monorepo** ç»“æ„ã€‚å‰ç«¯é‡‡ç”¨ **Feature-based** æ¶æ„ï¼Œåç«¯é‡‡ç”¨ **Modular** æ¶æ„ã€‚

```text
/
â”œâ”€â”€ apps
â”‚   â”œâ”€â”€ web  (Frontend: React + Vite)
â”‚   â”‚   â”œâ”€â”€ src
â”‚   â”‚   â”‚   â”œâ”€â”€ components      # ğŸ§© é€šç”¨åŸå­ç»„ä»¶ (Button, Inputå°è£…)
â”‚   â”‚   â”‚   â”œâ”€â”€ features        # ğŸ“¦ ä¸šåŠ¡æ¨¡å— (æ ¸å¿ƒæ¶æ„)
â”‚   â”‚   â”‚   â”‚   â””â”€â”€ [feature]   # e.g., "auth", "users"
â”‚   â”‚   â”‚   â”‚       â”œâ”€â”€ api.ts          # è¯¥æ¨¡å—çš„ React-Query hooks
â”‚   â”‚   â”‚   â”‚       â”œâ”€â”€ components/     # è¯¥æ¨¡å—ç‹¬æœ‰çš„ UI ç»„ä»¶
â”‚   â”‚   â”‚   â”‚       â”œâ”€â”€ routes/         # è¯¥æ¨¡å—çš„è·¯ç”±å®šä¹‰
â”‚   â”‚   â”‚   â”‚       â””â”€â”€ types.ts        # ä»…é™å‰ç«¯ä½¿ç”¨çš„ç±»å‹
â”‚   â”‚   â”‚   â”œâ”€â”€ hooks           # ğŸª é€šç”¨ Hooks (useDebounce, etc.)
â”‚   â”‚   â”‚   â”œâ”€â”€ layouts         # ğŸ–¼ï¸ å…¨å±€å¸ƒå±€ (Sidebar, Header)
â”‚   â”‚   â”‚   â”œâ”€â”€ providers       # ğŸ›¡ï¸ React Context Providers
â”‚   â”‚   â”‚   â”œâ”€â”€ routes          # ğŸš¦ è·¯ç”±å…¥å£
â”‚   â”‚   â”‚   â””â”€â”€ theme           # ğŸ¨ Ant Design Token é…ç½® (themeConfig.ts)
â”‚   â”‚   â””â”€â”€ package.json
â”‚   â”‚
â”‚   â””â”€â”€ api  (Backend: NestJS)
â”‚       â”œâ”€â”€ src
â”‚       â”‚   â”œâ”€â”€ common          # ğŸŒ å…¨å±€å®ˆå«, æ‹¦æˆªå™¨, è¿‡æ»¤å™¨
â”‚       â”‚   â”œâ”€â”€ config          # âš™ï¸ ç¯å¢ƒå˜é‡é…ç½®
â”‚       â”‚   â”œâ”€â”€ prisma          # ğŸ’¾ å…¨å±€ PrismaModule (@Global)
â”‚       â”‚   â”‚   â”œâ”€â”€ prisma.module.ts
â”‚       â”‚   â”‚   â”œâ”€â”€ prisma.service.ts
â”‚       â”‚   â”‚   â””â”€â”€ index.ts
â”‚       â”‚   â”œâ”€â”€ modules         # ğŸ§± ä¸šåŠ¡æ¨¡å— (ä¸€ä¸ªåŠŸèƒ½ = ä¸€ä¸ªæ¨¡å—)
â”‚       â”‚   â”‚   â””â”€â”€ [module]    # e.g., "users", "market-category"
â”‚       â”‚   â”‚       â”œâ”€â”€ dto/            # ä½¿ç”¨ createZodDto å°è£…
â”‚       â”‚   â”‚       â”œâ”€â”€ *.controller.ts # è·¯ç”±å¤„ç†
â”‚       â”‚   â”‚       â”œâ”€â”€ *.service.ts    # ä¸šåŠ¡é€»è¾‘
â”‚       â”‚   â”‚       â”œâ”€â”€ *.module.ts     # ä¾èµ–æ³¨å…¥
â”‚       â”‚   â”‚       â””â”€â”€ index.ts        # æ¨¡å—å¯¼å‡º
â”‚       â”‚   â””â”€â”€ main.ts         # å…¥å£æ–‡ä»¶
â”‚       â”œâ”€â”€ prisma              # Prisma Schema ç›®å½•
â”‚       â”‚   â””â”€â”€ schema.prisma
â”‚       â””â”€â”€ package.json
â”‚
â”œâ”€â”€ packages
â”‚   â”œâ”€â”€ types                   # ğŸ¤ã€æ ¸å¿ƒã€‘å‰åç«¯å…±äº«çš„ Zod Schemas & TS Types
â”‚   â”œâ”€â”€ tsconfig                # ğŸ› ï¸ å…±äº« TS é…ç½®
â”‚   â””â”€â”€ eslint-config           # ğŸ§¹ å…±äº« Lint é…ç½®
â”‚
â”œâ”€â”€ package.json (Root)
â””â”€â”€ pnpm-workspace.yaml
---

## 3. ğŸ›¡ï¸ å…¨å±€å¼€å‘æ³•åˆ™ (Global Rules)

### 3.1 ä¾èµ–ç®¡ç†
*   **PM**: ä¸¥ç¦ä½¿ç”¨ `npm` æˆ– `yarn`ã€‚**å¿…é¡»ä½¿ç”¨ `pnpm`**ã€‚
*   **å®‰è£…**:
    *   æ ¹ç›®å½•/å·¥å…·åº“: `pnpm add -w -D <pkg>`
    *   å‰ç«¯: `pnpm --filter web add <pkg>`
    *   åç«¯: `pnpm --filter api add <pkg>`
*   **æ— éšå¼ä¾èµ–**: æ‰€æœ‰ä½¿ç”¨çš„åŒ…å¿…é¡»åœ¨å¯¹åº”çš„ `package.json` ä¸­æ˜¾å¼å£°æ˜ã€‚

### 3.2 TypeScript é“å¾‹
*   **No Any**: ä¸¥ç¦ä½¿ç”¨ `any`ã€‚ä½¿ç”¨ `unknown` æˆ–å…·ä½“çš„ Interfaceã€‚
*   **å•ä¸€äº‹å®æ¥æº**:
    *   æ‰€æœ‰è·¨ç«¯æ•°æ®ç»“æ„ï¼ˆUser, Product, Orderï¼‰å¿…é¡»åœ¨ `packages/types` ä¸­å®šä¹‰ã€‚
    *   åç«¯è´Ÿè´£å®šä¹‰ Zod Schemaï¼Œå¯¼å‡º TypeScript ç±»å‹ç»™å‰ç«¯ä½¿ç”¨ã€‚

---

## 4. ğŸ–¥ï¸ å‰ç«¯æ³•åˆ™ (Frontend: apps/web)

### 4.1 æ ·å¼ä¸ UI (The "No CSS" Rule)
*   **Token ä¼˜å…ˆ**: ä¸¥ç¦ç¡¬ç¼–ç é¢œè‰²ï¼ˆå¦‚ `#1890ff`ï¼‰ã€‚å¿…é¡»ä½¿ç”¨ `theme.useToken()`ã€‚
*   **å¸ƒå±€ç»„ä»¶**: ä½¿ç”¨ `<Flex>`, `<Space>`, `<Row>`, `<Col>` ä»£æ›¿ CSS å¸ƒå±€ã€‚
*   **ç¦æ­¢å…¨å±€ CSS**: ä»…å…è®¸åœ¨ `App.css` ä¸­é‡ç½®æå°‘é‡çš„åŸºç¡€æ ·å¼ã€‚ç»„ä»¶æ ·å¼å¿…é¡»é€šè¿‡ `style={{ ... }}` (é…åˆ Token) æˆ– `emotion`/`css-modules` è§£å†³ã€‚

### 4.2 ç»„ä»¶æ„å»º
*   **ProComponents**:
    *   ç®¡ç†åå°çš„**è¡¨æ ¼**å¿…é¡»ç”¨ `ProTable`ã€‚
    *   ç®¡ç†åå°çš„**è¡¨å•**å¿…é¡»ç”¨ `ProForm` / `ModalForm` / `DrawerForm`ã€‚
    *   ç¦æ­¢æ‰‹å†™ Filter, Pagination, Loading é€»è¾‘ã€‚
*   **Feature-Sliced**: ä¸šåŠ¡ç»„ä»¶å¿…é¡»æ”¾åœ¨ `src/features/<feature-name>` ä¸‹ï¼Œä¸è¦æŠŠæ‰€æœ‰ä¸œè¥¿éƒ½å †åœ¨å…¨å±€ `components` é‡Œã€‚

### 4.3 æ•°æ®è·å–
*   **TanStack Query**:
    *   âŒ ç¦æ­¢åœ¨ useEffect ä¸­æ‰‹åŠ¨ fetch æ•°æ®ã€‚
    *   âœ… å¿…é¡»å°è£…ä¸º Custom Hook (e.g., `useUserList`) ä½¿ç”¨ `useQuery` æˆ– `useMutation`ã€‚

---

## 5. âš™ï¸ åç«¯æ³•åˆ™ (Backend: apps/api)

### 5.1 æ¶æ„åˆ†å±‚
*   **Controller**:
    *   âŒ **ç¦æ­¢ä¸šåŠ¡é€»è¾‘**ã€‚
    *   âœ… ä»…è´Ÿè´£ï¼šéªŒè¯ DTO -> è°ƒç”¨ Service -> è¿”å›ç»“æœã€‚
*   **Service**:
    *   âœ… è´Ÿè´£æ‰€æœ‰ä¸šåŠ¡åˆ¤æ–­ã€æ•°æ®åº“äº¤äº’ã€å¼‚å¸¸æŠ›å‡ºã€‚
*   **Prisma**:
    *   âŒ ç¦æ­¢åœ¨ Controller å±‚ç›´æ¥è°ƒç”¨ `prisma.*`ã€‚

### 5.2 éªŒè¯ (Validation)
*   **Zod DTO**:
    *   æ‰€æœ‰ Controller çš„ Payload å¿…é¡»é€šè¿‡ Zod Schema éªŒè¯ï¼ˆä½¿ç”¨ `nestjs-zod`ï¼‰ã€‚
    *   Schema å°½é‡å¤ç”¨ `packages/types` ä¸­çš„å®šä¹‰ã€‚

### 5.3 é”™è¯¯å¤„ç†
*   **HttpException**:
    *   å¿…é¡»æŠ›å‡º NestJS æ ‡å‡†å¼‚å¸¸ (e.g., `new BadRequestException('Invalid ID')`)ã€‚
    *   âŒ ç¦æ­¢ `console.log` é”™è¯¯åä¸åšå¤„ç†ã€‚

### 5.4 æ¨¡å—æ‹†åˆ†åŸåˆ™
*   **é«˜å†…èš**: ä¸€ä¸ªæ¨¡å—åªè´Ÿè´£ä¸€ä¸ªä¸šåŠ¡é¢†åŸŸã€‚
*   **ä½è€¦åˆ**: æ¨¡å—é—´é€šè¿‡ `exports` æš´éœ² Serviceï¼Œç¦æ­¢ç›´æ¥å¯¼å…¥å…¶ä»–æ¨¡å—çš„å†…éƒ¨æ–‡ä»¶ã€‚
*   **PrismaModule ç‰¹ä¾‹**: ä½œä¸ºå…¨å±€åŸºç¡€è®¾æ–½ï¼Œæ”¾åœ¨ `src/prisma/` è€Œé `src/modules/`ã€‚

ç¤ºä¾‹ï¼š
*   âœ… `market-category/`, `market-tag/`, `market-info/` (åˆ†å¼€)
*   âŒ `market-info/` åŒ…å« category + tag + info (åˆå¹¶)
---

## 6. ğŸ“ å‘½åè§„èŒƒ (Naming Convention)

| ç±»å‹ | è§„åˆ™ | ç¤ºä¾‹ |
| :--- | :--- | :--- |
| **React ç»„ä»¶æ–‡ä»¶** | PascalCase | `UserProfile.tsx` |
| **NestJS æ–‡ä»¶** | kebab-case | `auth.controller.ts` |
| **æ™®é€š TS æ–‡ä»¶** | camelCase | `dateUtils.ts` |
| **ç›®å½• (React)** | camelCase | `features/auth` |
| **å˜é‡/å‡½æ•°** | camelCase | `getUserInfo` |
| **Interface/Type** | PascalCase | `UserResponse` |

---

éµå®ˆæ­¤è§„åˆ™æ˜¯ä»£ç åˆå¹¶çš„å‰æã€‚